### PR TITLE
Display question feature

### LINKED ISSUE #44: Displaying certain question page feature
This feature is for the page that opens after you click certain questions on the "Explore" page.
The page will display the question, and underneath all the existing answers.
Also, if this page is opened by the user who wrote the question he will have the options to delete and edit.

### Tasks:
1. Add the needed function/class to views.py
2. Add the route to the urls,py
3. Create the needed HTML file
4. Add the needed tests for the feature

### The "prototype" design: 
image

### LINKED ISSUE #69: Write tests for display-certain-question feature

### PR DESCRIPTION
**Pull request by liornoy**
This PR is for the display question feature.
it includes changes in the URLs, views, models, and a new HTML file for this feature.
also added CSS file for all the needed classes.

Also added tests to tests.py using Pytest

fix #44 
fix #69

### CODE REVIEW

File: home/templates/home/question_detail.html
Comment (danitLevi) [inline home/templates/home/question_detail.html:54]: Good job Lior !
I think it would look better to show the number of likes and  the number of dislikes (next to the thumbs up and thumbs down icons) ,instead of showing the subtraction between them

File: home/templates/home/question_detail.html
Comment (danitLevi) [inline home/templates/home/question_detail.html:56]: Good job Lior !
I think it would look better to show the number of likes and the number of dislikes (next to the thumbs up and thumbs down icons) ,instead of showing the subtraction between them

Comment (liornoy) [reply home/templates/home/question_detail.html:56]: thanks, at first I thought to show the subtraction because this number dictates the display order of the answers when clicking the "votes" button. Nevertheless, I think showing both likes and dislikes can work great too.
I'll create the template for it and we'll compare and decide what to go with.

**Comment by danitLevi** [comment]
I think it's a better practice to add the "status/in progress" label instead of writing "wip" in the PR headline.

**Comment by idokk1** [comment]
LGTM

File: home/models.py
Comment (ifireball) [inline home/models.py:91]: If you need a method that returns an information that i relevant for a particular questions oyu can adde a regular method instead of a class method, that way you can avoid the extra DB query. The same comment seem to be relevant to all the methods you've added here.

File: home/views.py
Comment (ifireball) [inline home/views.py:17]: I think this is not a good function design - it has many very different things included in it, think about how difficult it would be to write tests for it.

I think it should be broken apart and most stuff in it moved out to model functions.

File: home/views.py
Comment (ifireball) [inline home/views.py:14]: Not sure what is the usefulnes of having class-based views here, IMO those seem to introduce extra code builerplate without visible benefits.

File: home/templates/home/question_detail.html
Comment (ifireball) [inline home/templates/home/question_detail.html:6]: I'd recommend having a shared, static CSS file - its better for browser-side caching.

File: home/tests.py
Comment (ifireball) [inline home/tests.py:7]: You don't need to define a test class and certainly nit inherit from TestCase. This is how you use [unittest][1] while we instructed you to use [PyTest][2]

[1]: https://docs.python.org/3/library/unittest.html
[2]: https://docs.pytest.org/en/6.2.x/

**Review by ifireball — CHANGES_REQUESTED** [review]

File: home/templates/home/question_detail.html
Comment (liornoy) [reply home/templates/home/question_detail.html:6]: We planned to move all styles to shard CSS file after everyone will merge their templates, so won't be conflicts in the shared file

Comment (ifireball) [reply home/templates/home/question_detail.html:6]: You'll get code duplication that way... If some parts of your UI have separate, distinct styles, you can put them in separate CSS files.

Comment (liornoy) [reply home/templates/home/question_detail.html]: I moved all the CSS to different file

**Comment by liornoy** [comment]
@ifireball , @AviadP the PR is ready for re-review, please

**Comment by ifireball** [comment]
You need to rebase this PR - it has conflicts - please don't leave it up to me to tell you this, you can look at the PR screen and scroll down just like I can.

File: home/models.py
Comment (liornoy) [reply home/models.py]: > You need to rebase this PR - it has conflicts - please don't leave it up to me to tell you this, you can look at the PR screen and scroll down just like I can.

Sorry I didn't notice the new PR install Pytest was merged.
I fixed my PR and is ready to review

File: home/views.py
Comment (ifireball) [inline home/views.py:12]: Not seeing any tests for this view

File: home/views.py
Comment (ifireball) [inline home/views.py:29]: This can be a custom template function, though IIRC the templating engine already has nice functions for dealing with plural v.s. sigular value titles.

File: home/tests.py
Comment (ifireball) [inline home/tests.py:8]: Since you're going to end up with a very big tests file, perhaps you want to split it up ahead of time, or at least use classes to group related tests together.

**Review by ifireball — CHANGES_REQUESTED** [review]
see comments on the code - generally I think this PR has too much stuff in it, so it takes a lot of time to review

File: home/views.py
Comment (liornoy) [reply home/views.py:12]: I don't understand what you mean you don't see tests for this view.
Every test exists because I need it for the view:
* The answer feed function which queries the answers to display,
* The thumb up / down functions
* The question title I use in the view
* The HTTP status response
All of them don't have a purpose without the view display question I made.
What am I missing?

File: home/views.py
Comment (liornoy) [reply home/views.py:29]: Ok, fixed

File: home/tests.py
Comment (liornoy) [reply home/tests.py:8]: Classes were added to group-related tests.

File: home/tests.py
Comment (ifireball) [inline home/tests.py:79]: Use the `client` fixture that the `pytest-django` plugin provides for you. We've shown at least two examples of using it in class.

Also you should probably test more then just the HTTP status code, you should at least verify that the right template had bee used and the context passed to it contains the right contents

File: home/views.py
Comment (ifireball) [reply home/views.py:12]: I think the method you have above that checks the view through the client wasn't there the last time I looked

File: home/views.py
Comment (ifireball) [inline home/views.py:20]: why is this function still here? ignoring my comments about it will not make me forget them.

**Review by ifireball — CHANGES_REQUESTED** [review]
see comments in the code

File: home/views.py
Comment (liornoy) [reply home/views.py]: I didn't ignore it, I did change and moved some functions to the models.py. Seeing that it's still not good enough, I moved all the context-making login to models.

File: home/tests.py
Comment (liornoy) [reply home/tests.py]: OK fixed

File: home/views.py
Comment (liornoy) [reply home/views.py]: fixed

File: home/tests.py
Comment (liornoy) [reply home/tests.py]: fixed

**Comment by liornoy** [comment]
> see comments in the code
@ifireball @AviadP
Fixed all the comments and ready for review

File: home/views.py
Comment (liornoy) [reply home/views.py]: changed to FBV

File: home/tests.py
Comment (ifireball) [inline home/tests.py:191]: this is checking the code with itself, its a useless test

if you want to ensure that function A calls function B and returns a value that is based on the value returned from function B, you need to use a mockup of function B and a sentinel value.

Mockups and sentinel values an be created with the unitest.mock library, and be injected into the application via the `monkeypatch` fixture.

But since you have very tight cntrol over the test data in the system, you can avoid doing any of that and just compare the data in the context you got to a hard-wired set of expected values.

**Review by ifireball — CHANGES_REQUESTED** [review]

File: home/models.py
Comment (liornoy) [reply home/models.py]: Issue fixed, made it dictionary creation expression

File: home/tests.py
Comment (liornoy) [reply home/tests.py]: Fixed, I compare it now with a hard-wired set of expected values.
Please note that I use string.startswith function because the question date is created to Now() time.
I know it's problematic, but I will change it after we will have more test data.
I opened an issue: #84 for it.

**Review by ifireball — APPROVED** [review]

**Pull request merged**
